internal/olm/operator/registry/operator_installer.go

```diff
@@ -122,49 +124,83 @@ func (o OperatorInstaller) waitForCatalogSource(ctx context.Context, cs *v1alpha
-	copy(targetNamespaces, o.InstallMode.TargetNamespaces)
+func (o OperatorInstaller) ensureOperatorGroup(ctx context.Context) (*v1.OperatorGroup, error) {
 	// Check OperatorGroup existence, since we cannot create a second OperatorGroup in namespace.
 	og, ogFound, err := o.getOperatorGroup(ctx)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	// TODO: we may need to poll for status updates, since status.namespaces may not be updated immediately.
-	if ogFound {
-		// targetNamespaces will always be initialized, but the operator group's namespaces may not be
-		// (required for comparison).
-		if og.Status.Namespaces == nil {
-			og.Status.Namespaces = []string{}
-		}
-		// Simple check for OperatorGroup compatibility: if namespaces are not an exact match,
-		// the user must manage the resource themselves.
-		sort.Strings(og.Status.Namespaces)
-		sort.Strings(targetNamespaces)
-		if !reflect.DeepEqual(og.Status.Namespaces, targetNamespaces) {
-			msg := fmt.Sprintf("namespaces %+q do not match desired namespaces %+q", og.Status.Namespaces, targetNamespaces)
-			if og.GetName() == operator.SDKOperatorGroupName {
-				return fmt.Errorf("existing SDK-managed operator group's %s, "+
-					"please clean up existing operators `operator-sdk cleanup` before running package %q", msg, o.PackageName)
+
+	supported := o.SupportedInstallModes
+
+	if supported.Len() == 0 {
+		return nil, fmt.Errorf("operator %q is not installable: no supported install modes", o.StartingCSV)
+	}
+
+	// --install-mode was given
+	if !o.InstallMode.IsEmpty() {
+		// TODO: probably remove multinamespace
+		if o.InstallMode.InstallModeType == v1alpha1.InstallModeTypeSingleNamespace {
+			targetNsSet := sets.NewString(o.InstallMode.TargetNamespaces...)
+			if !supported.Has(string(v1alpha1.InstallModeTypeOwnNamespace)) && targetNsSet.Has(o.cfg.Namespace) {
+				return nil, fmt.Errorf("cannot watch namespace %q: operator %q does not support install mode %q", o.cfg.Namespace, o.StartingCSV, v1alpha1.InstallModeTypeOwnNamespace)
 			}
-			return fmt.Errorf("existing operator group %q's %s, "+
-				"please ensure it has the exact namespace set before running package %q", og.GetName(), msg, o.PackageName)
 		}
-		log.Infof("Using existing operator group %q", og.GetName())
-	} else {
-		// New SDK-managed OperatorGroup.
-		og = newSDKOperatorGroup(o.cfg.Namespace,
-			withTargetNamespaces(targetNamespaces...))
-		if err = o.cfg.Client.Create(ctx, og); err != nil {
-			return fmt.Errorf("error creating OperatorGroup: %w", err)
+		if o.InstallMode.InstallModeType == v1alpha1.InstallModeTypeSingleNamespace &&
+			o.InstallMode.TargetNamespaces[0] == o.cfg.Namespace {
+			return nil, fmt.Errorf("use install mode %q to watch operator's namespace %q", v1alpha1.InstallModeTypeOwnNamespace, o.cfg.Namespace)
+		}
+
+		supported = supported.Intersection(sets.NewString(string(o.InstallMode.InstallModeType)))
+		if supported.Len() == 0 {
```

Comment:
```suggestion
		if !supported.Has(string(o.InstallMode.InstallModeType)) {
```

Reply:
The intersection actually changes `supported`. Basically makes it the only one in the set. And `getTargetNamespaces` uses the supported set to determine what namespaces to return.